### PR TITLE
[Proposal] Orthogonal Initializer 

### PR DESCRIPTION
**Pull request by joaogui1**
This PR implements the [orthogonal intialization](https://smerity.com/articles/2016/orthogonal_init.html), useful for training RNNs (and maybe general Neural Nets when combined with Orthogonal regularization)
Is there interest in this PR?

### CODE REVIEW

**Comment by david-berthelot** [comment]
Definitely welcome, is it ready for review or is it in progress?
About the orthogonalization axis, what is its typical value (fanin, output, something else)?

**Comment by joaogui1** [comment]
My bad, the orthogonalization axis is generally the last one. The PR is ready to be reviewed @david-berthelot

File: objax/nn/init.py [hunk 1]
```diff
@@ -91,6 +91,41 @@ def kaiming_truncated_normal(shape: Tuple[int, ...], lower: float = -2, upper: f
...
+    matrix_shape = (n_rows, n_cols) if n_rows > n_cols else (n_cols, n_rows)
+    norm_dst = random.normal(matrix_shape)
+    q_mat, r_mat = np.linalg.qr(norm_dst)
+    # Enforce Q is uniformly distributed
+    q_mat *= np.sign(np.diag(r_mat))
+    if n_rows < n_cols:
+      q_mat = q_mat.T
+    q_mat = np.reshape(q_mat, (n_rows,) + tuple(np.delete(shape, axis)))
+    q_mat = np.moveaxis(q_mat, 0, axis)
+    return gain * q_mat
```
Comment (david-berthelot) [inline objax/nn/init.py:126]: Let's return a JAX array.
```suggestion
    return gain * jn.array(q_mat)
```

File: objax/nn/init.py [hunk 2]
```diff
@@ -156,4 +191,4 @@ def xavier_truncated_normal(shape: Tuple[int, ...], lower: float = -2, upper: fl
     """
     truncated_std = scipy.stats.truncnorm.std(a=lower, b=upper, loc=0., scale=1)
     stddev = gain * xavier_normal_gain(shape) / truncated_std
-    return random.truncated_normal(shape, stddev=stddev, lower=lower, upper=upper)
+    return random.truncated_normal(shape, stddev=stddev, lower=lower, upper=upper)
```
Comment (david-berthelot) [inline objax/nn/init.py:194]: I think you accidentally removed a newline character at the end.

**Review by david-berthelot — CHANGES_REQUESTED** [review]
Can you also add some unit tests in `tests/`, there are already some examples to get inspired from.

File: tests/nn_init.py [hunk 3]
```diff
@@ -91,11 +91,11 @@ def test_orthogonal(self):
         for s in ((10, 100), (100, 10), (10, 10)):
             for gain in (1, 2):
                 init = np.asarray(objax.nn.init.orthogonal(s, gain=gain))
-                I = init@init.T if s[0] < s[1] else init.T@init
-                I /= gain**2
-                diff = I - np.eye(np.min(s))
+                I_ = init @ init.T if s[0] < s[1] else init.T @ init
+                I_ /= gain**2
+                diff = I_ - np.eye(np.min(s))
```
Comment (david-berthelot) [inline tests/nn_init.py:96]: Can we use non-uppercase letters (they're typically for global constants)?

**Review by david-berthelot — CHANGES_REQUESTED** [review]

**Review by david-berthelot — APPROVED** [review]
Thanks!
